// react-pick-datetime/lib/CalendarGridBody.js
"use strict";

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var React = require("react/addons");

var PureRenderMixin = React.addons.PureRenderMixin;

var moment = require("moment");
var joinClasses = require("react/lib/joinClasses");

function getWeeks(month, value) {
  var today = moment();
  var endDay = moment(month).endOf("month").endOf("week");
  var currentDay = moment(month).startOf("month").startOf("week");
  var currentWeek;
  var weeks = [];

  while (!currentDay.isAfter(endDay)) {
    if (currentDay.day() === 0) {
      currentWeek = [];
      weeks.push(currentWeek);
    }

    currentWeek.push({
      isOtherMonth: currentDay.month() !== month.month(),
      isValue: value && currentDay.isSame(value, "day"),
      isToday: currentDay.isSame(today, "day"),
      value: moment(currentDay),
      formatted: currentDay.format("D")
    });

    currentDay.add(1, "days");
  }

  return weeks;
}

var CalendarGridBody = React.createClass({
  displayName: "CalendarGridBody",

  mixins: [PureRenderMixin],

  propTypes: {
    getDescendantIdForDay: React.PropTypes.func.isRequired,
    month: React.PropTypes.object.isRequired,
    onComplete: React.PropTypes.func.isRequired,
    value: React.PropTypes.object
  },

  getDefaultProps: function getDefaultProps() {
    return {
      value: null
    };
  },

  handleDayClick: function handleDayClick(day) {
    this.props.onComplete(day);
  },

  render: function render() {
    var _this = this;

    var _props = this.props;
    var month = _props.month;
    var value = _props.value;

    var otherProps = _objectWithoutProperties(_props, ["month", "value"]);

    var weeks = getWeeks(month, value);

    return React.createElement(
      "tbody",
      otherProps,
      weeks.map(function (week, idx) {
        return React.createElement(
          "tr",
          { key: idx },
          week.map(function (day, idx) {
            return React.createElement(
              "td",
              {
                "aria-selected": day.isSelected + "",
                className: joinClasses("CalendarGridBody-day", day.isToday && "CalendarGridBody-day--isToday", day.isOtherMonth && "CalendarGridBody-day--isOtherMonth", day.isValue && "CalendarGridBody-day--isValue"),
                id: _this.props.getDescendantIdForDay(day.value),
                key: idx,
                onClick: _this.handleDayClick.bind(_this, day.value),
                role: "gridcell" },
              day.formatted
            );
          })
        );
      })
    );
  }

});

module.exports = CalendarGridBody;
