// react-pick-datetime/lib/CalendarGridHeader.js
"use strict";

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var React = require("react/addons");

var PureRenderMixin = React.addons.PureRenderMixin;

var moment = require("moment");

function getDayHeadings(month) {
  var endDay = moment(month).endOf("week");
  var currentDay = moment(month).startOf("week");
  var headings = [];

  while (!currentDay.isAfter(endDay)) {
    headings.push({
      abbr: currentDay.format("dd"),
      title: currentDay.format("dddd")
    });

    currentDay.add(1, "days");
  }

  return headings;
}

var CalendarGridHeader = React.createClass({
  displayName: "CalendarGridHeader",

  mixins: [PureRenderMixin],

  propTypes: {
    month: React.PropTypes.object.isRequired
  },

  render: function render() {
    var _props = this.props;
    var month = _props.month;

    var otherProps = _objectWithoutProperties(_props, ["month"]);

    var headings = getDayHeadings(month);

    return React.createElement(
      "thead",
      otherProps,
      React.createElement(
        "tr",
        null,
        headings.map(function (heading, idx) {
          return React.createElement(
            "th",
            { key: idx },
            React.createElement(
              "abbr",
              { tile: heading.title },
              heading.abbr
            )
          );
        })
      )
    );
  }

});

module.exports = CalendarGridHeader;
