// react-pick-datetime/lib/DateInput.js
"use strict";

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var DateInput = require("./DateInput");
var DateInputButton = require("./DateInputButton");
var DatePopup = require("./DatePopup");
var React = require("react/addons");

var _require = require("react-pick");

var InputPopupWrapper = _require.InputPopupWrapper;
var PureRenderMixin = React.addons.PureRenderMixin;

var moment = require("moment");
var joinClasses = require("react/lib/joinClasses");

var DateInput = React.createClass({
  displayName: "DateInput",

  mixins: [PureRenderMixin],

  propTypes: {
    /**
     * Event handler fired when the value of the `<DateInput>` changes.
     * The called function is passed `value`.
     */
    onChange: React.PropTypes.func.isRequired,

    /**
     * The current value of the `<DateInput>`, as a `moment` date.
     */
    value: React.PropTypes.object,

    /**
     * The format of the date shown and parsed in the `<input> box.
     * Default is `'l'`.
     */
    inputValueFormat: React.PropTypes.string,

    /**
     * The component to render for the input.
     * Default is `input`.
     */
    inputComponent: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes["function"]]),

    /**
     * The component to render for the button that lets you toggle the popup.
     * Default is `DateInputButton`.
     */
    buttonComponent: React.PropTypes.func,

    /**
     * The component to render for the date popup.
     * Default is `DatePopup`.
     */
    popupComponent: React.PropTypes.func
  },

  getDefaultProps: function getDefaultProps() {
    return {
      value: null,
      inputValueFormat: "l",
      inputComponent: "input",
      buttonComponent: DateInputButton,
      popupComponent: DatePopup
    };
  },

  getInitialState: function getInitialState() {
    var _props = this.props;
    var value = _props.value;
    var inputValueFormat = _props.inputValueFormat;

    return {
      isOpen: false,
      popupMonth: value || moment(),
      inputValue: value !== null ? value.format(inputValueFormat) : null
    };
  },

  handlePopupMonthChange: function handlePopupMonthChange(popupMonth) {
    this.setState({ popupMonth: popupMonth });
  },

  handlePopupChange: function handlePopupChange(value) {
    this.setState({
      inputValue: value.format(this.props.inputValueFormat),
      popupMonth: value || this.state.popupMonth
    });

    this.props.onChange(value);
  },

  handlePopupComplete: function handlePopupComplete(value) {
    this.handlePopupChange(value);
    this.setState({ isOpen: false });
  },

  handlePopupCancel: function handlePopupCancel() {
    this.setState({ isOpen: false });
  },

  handleInputChange: function handleInputChange(event) {
    var inputValue = event.target.value;
    var _props = this.props;
    var inputValueFormat = _props.inputValueFormat;
    var onChange = _props.onChange;

    this.setState({ inputValue: inputValue });

    var parsedInputValue = moment(inputValue, inputValueFormat, true);
    if (parsedInputValue.isValid()) {
      this.setState({ popupMonth: parsedInputValue });
      onChange(parsedInputValue);
    } else {
      onChange(null);
    }
  },

  handleButtonClick: function handleButtonClick() {
    var _this = this;

    this.setState({ isOpen: !this.state.isOpen }, function () {
      _this.state.isOpen && _this.refs.popup.focusOnGrid();
    });
  },

  renderPopup: function renderPopup() {
    var PopupComponent = this.props.popupComponent;

    return React.createElement(PopupComponent, {
      ref: "popup",
      month: this.state.popupMonth,
      value: this.props.value,
      onMonthChange: this.handlePopupMonthChange,
      onChange: this.handlePopupChange,
      onComplete: this.handlePopupComplete,
      onCancel: this.handlePopupCancel
    });
  },

  render: function render() {
    var ButtonComponent = this.props.buttonComponent;
    var InputComponent = this.props.inputComponent;
    var _props = this.props;
    var className = _props.className;

    var otherProps = _objectWithoutProperties(_props, ["className"]);

    return React.createElement(
      "div",
      { className: joinClasses("DateInput", className) },
      React.createElement(
        InputPopupWrapper,
        {
          isOpen: this.state.isOpen,
          popupElement: this.renderPopup() },
        React.createElement(InputComponent, _extends({}, otherProps, {
          value: this.state.inputValue,
          onFocus: this.handleInputFocus,
          onBlur: this.handleInputBlur,
          onChange: this.handleInputChange
        }))
      ),
      React.createElement(ButtonComponent, {
        "aria-hidden": "true",
        onClick: this.handleButtonClick
      })
    );
  }

});

module.exports = DateInput;
