// react-pick-datetime/lib/DatePopupHeader.js
"use strict";

var React = require("react/addons");

var PureRenderMixin = React.addons.PureRenderMixin;

var moment = require("moment");

var DatePopupHeader = React.createClass({
  displayName: "DatePopupHeader",

  mixins: [PureRenderMixin],

  propTypes: {
    getHeadingId: React.PropTypes.func.isRequired,
    onMonthChange: React.PropTypes.func.isRequired,
    month: React.PropTypes.object.isRequired
  },

  handleNextButtonClick: function handleNextButtonClick() {
    this.props.onMonthChange(moment(this.props.month).add(1, "months"));
  },

  handlePreviousButtonClick: function handlePreviousButtonClick() {
    this.props.onMonthChange(moment(this.props.month).subtract(1, "months"));
  },

  render: function render() {
    var _props = this.props;
    var getHeadingId = _props.getHeadingId;
    var month = _props.month;

    return React.createElement(
      "div",
      { className: "DatePopupHeader" },
      React.createElement(
        "button",
        {
          className: "DatePopupHeader-previousButton",
          onClick: this.handlePreviousButtonClick,
          tabIndex: "0" },
        "◀"
      ),
      React.createElement(
        "button",
        {
          className: "DatePopupHeader-nextButton",
          onClick: this.handleNextButtonClick,
          tabIndex: "0" },
        "▶"
      ),
      React.createElement(
        "div",
        {
          "aria-live": "assertive",
          "aria-atomic": "true",
          className: "DatePopupHeader-heading",
          id: getHeadingId(),
          role: "heading" },
        month.format("MMMM YYYY")
      )
    );
  }

});

module.exports = DatePopupHeader;
