// react-pick-datetime/lib/DatePopupGridHeader.js
"use strict";

var React = require("react/addons");

var PureRenderMixin = React.addons.PureRenderMixin;

var moment = require("moment");

function getDayHeadings(month) {
  var endDay = moment(month).endOf("week");
  var currentDay = moment(month).startOf("week");
  var headings = [];

  while (!currentDay.isAfter(endDay)) {
    headings.push({
      abbr: currentDay.format("dd"),
      title: currentDay.format("dddd")
    });

    currentDay.add(1, "days");
  }

  return headings;
}

var DatePopupGridHeader = React.createClass({
  displayName: "DatePopupGridHeader",

  mixins: [PureRenderMixin],

  propTypes: {
    month: React.PropTypes.object.isRequired
  },

  render: function render() {
    var headings = getDayHeadings(this.props.month);

    return React.createElement(
      "thead",
      null,
      React.createElement(
        "tr",
        null,
        headings.map(function (heading, idx) {
          return React.createElement(
            "th",
            { key: idx },
            React.createElement(
              "abbr",
              { tile: heading.title },
              heading.abbr
            )
          );
        })
      )
    );
  }

});

module.exports = DatePopupGridHeader;
